### PR TITLE
Editor sticky notes

### PR DESCRIPTION
**Pull request by andreas-schultz**

### CODE REVIEW

File: workspace/src/app/views/shared/RuleEditor/model/RuleEditorModel.tsx [hunk 1]
```diff
@@ -1407,7 +1647,15 @@ export const RuleEditorModel = ({ children }: RuleEditorModelProps) => {
                 output: targetNode.get(opNode.nodeId),
             })
         );
-        let elems: Elements = [...nodes, ...edges];
+
+        const stickyNodes = ruleEditorContext.stickyNotes.map(({ color, content, position }) =>
+            createStickyNodeInternal(color, content, {
```
Comment (andreas-schultz) [inline workspace/src/app/views/shared/RuleEditor/model/RuleEditorModel.tsx:1652]: Use the ID of the sticky notes coming from the backend.

File: workspace/src/app/views/shared/RuleEditor/model/RuleEditorModel.tsx [hunk 2]
```diff
@@ -1422,6 +1670,47 @@ export const RuleEditorModel = ({ children }: RuleEditorModelProps) => {
         }, 1);
     };
 
+    const addStickyNoteToCanvas = (stickyNote: string, color: string, reactFlowWrapper) => {
```
Comment (andreas-schultz) [inline workspace/src/app/views/shared/RuleEditor/model/RuleEditorModel.tsx:1673]: There are several things that are not clear to me:
- This is called `addStickyNoteToCanvas`, but it also seems to update existing sticky notes. Mabye there should be 2 functions instead?
- Why is `currentStickyContent` used, e.g. why not use an optional node ID as parameter?
- Both functions ` changeStickyNodeStyle` and `changeStickyNoteContent` are both part of the context, but only used internally in the rule editor model.

File: workspace/src/app/views/shared/RuleEditor/model/RuleEditorModel.tsx [hunk 3]
```diff
@@ -1422,6 +1670,47 @@ export const RuleEditorModel = ({ children }: RuleEditorModelProps) => {
...
+                y: (reactFlowBounds.height - DEFAULT_NODE_HEIGHT) / 2,
+            });
+            addStickyNode(stickyNote, position, color);
+        } else {
+            changeStickyNodeStyle(nodeId, color);
+            changeStickyNoteContent(nodeId, stickyNote);
+        }
+    };
+
+    function stickyMenuButtons(stickyId, color, stickyNote) {
```
Comment (andreas-schultz) [inline workspace/src/app/views/shared/RuleEditor/model/RuleEditorModel.tsx:1688]: Move rendering code into utils file.

File: workspace/src/app/views/shared/RuleEditor/model/RuleEditorModel.tsx [hunk 4]
```diff
@@ -1422,6 +1670,47 @@ export const RuleEditorModel = ({ children }: RuleEditorModelProps) => {
...
+        return (
+            <>
+                <IconButton
+                    data-test-id={"edit-sticky-note"}
+                    name="item-edit"
+                    text={t("RuleEditor.node.executionButtons.edit.tooltip")}
+                    onClick={() => {
+                        setCurrentStickyContent(
```
Comment (andreas-schultz) [inline workspace/src/app/views/shared/RuleEditor/model/RuleEditorModel.tsx:1696]: As already mentioned above the "view" code for editing the sticky note should probably be somewhere else, e.g. a `StickyNoteMenu` component.

File: workspace/src/app/views/shared/RuleEditor/model/test/RuleEditorModel.test.tsx [hunk 5]
```diff
@@ -286,6 +297,107 @@ describe("Rule editor model", () => {
...
+            expect(modelUtils.nodeById(currentContext().elements, node.id)!!.data.style).not.toStrictEqual(
+                defaultStyle
+            );
+        };
+        act(() => {
+            currentContext().executeModelEditOperation.changeStickyNodeStyle(node.id, "#fee2f1");
+        });
+        assertionAfterSizeChange();
+
+        //UNDO
```
Comment (andreas-schultz) [inline workspace/src/app/views/shared/RuleEditor/model/test/RuleEditorModel.test.tsx:323]: Why not use `checkUndoAndRedo`? Same for the other tests that were added.

File: workspace/src/app/views/shared/RuleEditor/model/test/RuleEditorModel.test.tsx [hunk 6]
```diff
@@ -702,18 +815,39 @@ describe("Rule editor model", () => {
...
             console.log("Test UNDO");
             for (let changeIdx = stateHistory.length - 1; changeIdx > 0; changeIdx--) {
                 expect(currentContext().canUndo).toBe(true);
                 act(() => {
                     currentContext().undo();
                 });
-                console.log(`Undone change: ${stateHistoryLabel[changeIdx]} (${changeIdx}/${stateHistory.length - 1})`);
-                expect(currentOperatorNodes()).not.toStrictEqual(stateHistory[changeIdx]);
-                expect(currentOperatorNodes()).toStrictEqual(stateHistory[changeIdx - 1]);
+                const change = stateHistory[changeIdx];
```
Comment (andreas-schultz) [inline workspace/src/app/views/shared/RuleEditor/model/test/RuleEditorModel.test.tsx:844]: Why is this condition needed here?

File: workspace/src/app/views/shared/RuleEditor/view/RuleEditorCanvas.tsx [hunk 7]
```diff
@@ -500,56 +509,63 @@ export const RuleEditorCanvas = () => {
...
-                    nodeClassName={minimapNodeClassName}
-                    nodeColor={minimapNodeColor}
-                    enableNavigation={true}
-                />
-                <Controls
-                    showInteractive={!!modelContext.setIsReadOnly}
-                    onInteractiveChange={(isInteractive) =>
-                        modelContext.setIsReadOnly && modelContext.setIsReadOnly(!isInteractive)
+        <>
+            {modelContext.showStickyNoteModal ? (
```
Comment (andreas-schultz) [inline workspace/src/app/views/shared/RuleEditor/view/RuleEditorCanvas.tsx:513]: As noted before, the sticky note modal can be much more "local" where the actual create/update action is fired.

File: workspace/src/app/views/taskViews/linking/LinkingRuleEditor.tsx [hunk 8]
```diff
@@ -149,6 +153,9 @@ export const LinkingRuleEditor = ({ projectId, linkingTaskId, viewActions }: Lin
                 ...originalRule,
                 operator: ruleTree,
                 layout: ruleUtils.ruleLayout(ruleOperatorNodes),
+                uiAnnotations: {
+                    stickyNotes, // TODO: Add sticky notes. Also read sticky notes from backend.
```
Comment (andreas-schultz) [inline workspace/src/app/views/taskViews/linking/LinkingRuleEditor.tsx:157]: Remove TODO

File: workspace/src/app/views/shared/RuleEditor/contexts/RuleEditorModelContext.ts [hunk 9]
```diff
@@ -42,11 +44,17 @@ export interface RuleEditorModelContextProps {
     centerNode: (nodeId: string) => boolean;
     /** Get the current rule as IRuleOperatorNode objects. */
     ruleOperatorNodes: () => IRuleOperatorNode[];
+    allStickyNodes: () => IStickyNote[];
```
[the remark this answers is not in the document]
Comment (arausly) [reply workspace/src/app/views/shared/RuleEditor/contexts/RuleEditorModelContext.ts:47]: it is used also in the save function apart from the test

File: workspace/src/app/views/shared/RuleEditor/contexts/RuleEditorModelContext.ts [hunk 10]
```diff
@@ -42,11 +44,17 @@ export interface RuleEditorModelContextProps {
...
+    currentStickyContent: Map<string, string>;
+    showStickyNoteModal: boolean;
```
[the remark this answers is not in the document]
Comment (arausly) [reply workspace/src/app/views/shared/RuleEditor/contexts/RuleEditorModelContext.ts:49]: it is also used in the ruleEditorModel, the alternative would be to raise the state higher in context tree like the modal in the toolbar, which would be more work than necessary in my opinion. I might be wrong :)

**Pull request merged**